### ISSUE TITLE
is it possible to return different versions of views based on authenticated or not

### ISSUE THREAD

**Issue by fyears** [miguelgrinberg/Flask-HTTPAuth#17]
For example, I am building a twitter-like website. I would like the website has these responses:

```bash
curl https://localhost/home
# => return a public timeline
curl -u username:password https://localhost/home
# => return the personal timeline
```

However, if I write code like this:
```python
@app.route('/home')
@auth.login_required
def home():
   pass
```

I would get:
```bash
curl https://localhost/home
# => 401 error (default in Flask-HTTPAuth)
```

**Comment by miguelgrinberg** [comment]
This is currently not possible, but you are not the first to request this. The problem is that the current version of the extension will directly return a 401 error when the request does not include the authorization header. The change that I was planning to make was to allow the password validation password to be called even when the header is missing. You can then return `True` from the password validation callback and set something in `flask.g` to indicate to the route that there is no user authenticated.

**Comment by fyears** [comment]
But still, it seems that your solution right now still requires an auth header in the request, or it would abort 401 anyway.

**Comment by miguelgrinberg** [comment]
Correct, this requires a change that I was considering for the future, it
is not possible right now. I'll try to look into it tomorrow.

**Comment by miguelgrinberg** [comment]
Addressed in version 2.4.0. In this version for requests that don't have an `Authorization` header the extension calls the `verify_password()` callback with `username` and `password` set to empty strings. You can opt to return `True` and that will let the request through. You can write something to `flask.g` so that the route know the user is anonymous.

**Comment by fyears** [comment]
Thank you.

Just a reference, this issue is solved in commit 5c5396bbb7af540a7aff786ce3282657566045f2 .
